### PR TITLE
chore: add `--debug-with-logfire` to pytest

### PR DESCRIPTION
**Pull request by dsfaccini**
we have a barebones logfire fixture, I've extended it so it can be activated with a flag `--debug-with-logfire`, it creates a single span for the `pytest` command and wraps each single test in its own span

image

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -279,9 +279,25 @@ def no_instrumentation_by_default():
...
     @pytest.fixture(autouse=True)
-    def fresh_logfire():
-        logfire.shutdown(flush=False)
+    async def fresh_logfire(request: pytest.FixtureRequest, _logfire_session: None, anyio_backend: str):
+        if request.config.getoption('debug_with_logfire'):
+            with logfire.span('test: {test_name}', test_name=request.node.name):
+                yield
+        else:
+            logfire.shutdown(flush=False)
+            yield
```
[the remark this answers is not in the document]
Comment (dsfaccini) [reply tests/conftest.py:300]: couldn't confirm after running various tests with the flag

**Comment by dsfaccini** [comment]
one edge case is that running `tests/test_logfire.py` with the flag will produce snapshot mismatches, but that shouldn't be a concern. debugging with logfire is meant for debugging model provider API tests

File: tests/conftest.py [hunk 2]
```diff
@@ -279,9 +279,25 @@ def no_instrumentation_by_default():
 
     logfire.DEFAULT_LOGFIRE_INSTANCE.config.ignore_no_config = True
 
+    @pytest.fixture(scope='session')
+    def _logfire_session(request: pytest.FixtureRequest):
+        if request.config.getoption('debug_with_logfire'):
+            logfire.configure(send_to_logfire='if-token-present', console=False)
+            logfire.instrument_httpx(capture_all=True)
```
Comment (DouweM) [inline tests/conftest.py:286]: No `instrument_pydantic_ai`?

File: tests/conftest.py [hunk 3]
```diff
@@ -312,6 +328,13 @@ def pytest_addoption(parser: Any) -> None:
         dest='xai_proto_include_json',
         help='Include JSON representations in xAI proto cassette YAML files.',
     )
+    parser.addoption(
+        '--debug-with-logfire',
```
Comment (DouweM) [inline tests/conftest.py:332]: Why not just `--logfire`?

File: tests/conftest.py [hunk 4]
```diff
@@ -331,14 +354,18 @@ def mock_vcr_aiohttp_content(mocker: MockerFixture):
...
-    return {
+@pytest.fixture(scope='function')
+def vcr_config(request: pytest.FixtureRequest):
+    config: dict[str, Any] = {
         'ignore_localhost': True,
         # Note: additional header filtering is done inside the serializer
         'filter_headers': ['authorization', 'x-api-key'],
         'decode_compressed_response': True,
     }
+    if request.config.getoption('debug_with_logfire'):
```
Comment (DouweM) [inline tests/conftest.py:365]: Should have a constant for this string

**Review by DouweM — CHANGES_REQUESTED** [review]

**Comment by DouweM** [comment]
> one edge case is that running tests/test_logfire.py with the flag will produce snapshot mismatches, but that shouldn't be a concern. debugging with logfire is meant for debugging model provider API tests

@dsfaccini Could we have a fixture that we use in `test_logfire` to skip the extra layer of instrumentation?

File: tests/conftest.py (hunk 3 above)
Comment (dsfaccini) [reply tests/conftest.py:332]: I prefer explicit flags and I lowkey didn't want llms to confuse `--lf` with `--logfire`

File: tests/conftest.py (hunk 2 above)
Comment (dsfaccini) [reply tests/conftest.py:286]: when I'm looking for errors I look into the HTTP, which is why I thought default httpx instrumentation was better, but I'm not stubborn on this, lmk which default you want and whether we should have separate flags like `--instrument-pai` and `--instrument-httpx`

**Comment by dsfaccini** [comment]
I'll look into it

File: tests/conftest.py (hunk 2 above)
Comment (DouweM) [reply tests/conftest.py:286]: I always prefer to instrument both, just gives more context into what's happening with tool calls etc

File: tests/conftest.py (hunk 3 above)
Comment (DouweM) [reply tests/conftest.py:332]: But all test running is basically debugging right? I think `--instrument` is at least as explicit

**Comment by Kludex** [comment]
We have a proper integration in Logfire now: https://pydantic.dev/docs/logfire/integrations/pytest

I'll close this as stale, since it's not using that approach. Please discuss between yourselves if you want to add that pytest plugin to the repository.
